framework/include/userobject/DomainUserObject.h

```diff
@@ -0,0 +1,235 @@
+  void preExecuteOnInternalSide();
+
+  /**
+   * method that is called right before executeOnInterface; sets the data to face
+   */
+  void preExecuteOnInterface();
+
+  /**
+   * Return whether this object should run \p executeOnInterface
+   */
+  bool shouldExecuteOnInterface() const;
+
+  void checkVariable(const MooseVariableFieldBase & variable) const override;
+
+protected:
+  const MooseArray<Point> & qPoints() const { return *_current_q_point; }
+  const QBase & qRule() const { return *_current_q_rule; }
+  const MooseArray<Real> & JxW() const { return *_current_JxW; }
+  const MooseArray<Real> & coord() const { return _coord; }
+  const MooseArray<Point> & normals() const { return _normals; }
+
+  /**
+   * Routes through to \p Coupleable::getFieldVar, but also inserts the return variable into a set
+   * of field variables to check on interface-connected blocks, as opposed to our blocks, when
+   * performing our block-restriction integrity check
+   */
+  const MooseVariableFieldBase * getInterfaceFieldVar(const std::string & var_name,
+                                                      unsigned int comp);
+
+  /// the Moose mesh
+  MooseMesh & _mesh;
+
+  /// The current element pointer (available during all execute functions)
+  const Elem * const & _current_elem;
+
+  /// The current element volume (available during all execute functions)
+  const Real & _current_elem_volume;
+
+  /// Current side of the current element (available during executeOnInternalSide() and
+  /// executeOnBoundary() and executeOnInterface())
+  const unsigned int & _current_side;
+
+  /// Current side of the current element (available during executeOnInternalSide() and
+  /// executeOnBoundary() and executeOnInterface())
+  const Elem * const & _current_side_elem;
+
+  /// Current side volume (available during executeOnInternalSide() and executeOnBoundary() and
+  /// executeOnInterface())
+  const Real & _current_side_volume;
+
+  /// The neighboring element (available during executeOnInternalSide() and executeOnInterface())
+  const Elem * const & _neighbor_elem;
+
+  /// the neighboring element's volume (available during executeOnInternalSide() and
+  /// executeOnInterface())
+  const Real & _current_neighbor_volume;
+
+  /// The boundary ID (available during executeOnBoundary() and executeOnInterface())
+  const BoundaryID & _current_boundary_id;
+
+  /// The unit norm at quadrature points on the element side/face from the current element
+  /// perpendicular to the side
+  const MooseArray<Point> & _normals;
+
+  /// The set of boundary IDs on which this object should perform \p executeOnInterface
+  std::set<BoundaryID> _interface_bnd_ids;
+
+  /// The set of blocks connected to our blocks through the \p _interface_bnd_ids data member
+  std::set<SubdomainID> _interface_connected_blocks;
+
+  /// The set of variables we wish to evaluate on the interface connected blocks
+  std::set<const MooseVariableFieldBase *> _interface_vars;
+
+private:
```

Comment:
I'm glad we're keeping this all private; there's so much room for error with everything combined in an object